src/merkle/partial_mt/mod.rs

```diff
@@ -327,3 +347,64 @@ impl PartialMerkleTree {
         Ok(())
     }
 }
+
+// SERIALIZATION
+// ================================================================================================
+
+impl Serializable for PartialMerkleTree {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        // write max_depth
+        target.write_u8(self.max_depth);
+
+        // write leaf nodes
+        target.write_u64(self.leaves.len() as u64);
+        for leaf_index in self.leaves.iter() {
+            leaf_index.write_into(target);
+            self.get_node(*leaf_index).expect("Leaf hash not found").write_into(target);
+        }
+    }
+}
+
+impl Deserializable for PartialMerkleTree {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        // get max_depth
+        let max_depth = source.read_u8()?;
+
+        // add leaf nodes to the map
+        let mut nodes = BTreeMap::new();
+        let leaves_len = source.read_u64()?;
+        for _ in 0..leaves_len {
+            let index = NodeIndex::read_from(source)?;
+            let hash = RpoDigest::read_from(source)?;
+            nodes.insert(index, hash);
+        }
+
+        // get leaves set
+        let leaves = nodes.keys().copied().collect();
+
+        // calculate and add internal nodes
+        for layer_number in (1..max_depth + 1).rev() {
+            let layer: BTreeMap<NodeIndex, RpoDigest> = nodes
+                .iter()
+                .filter(|(index, _)| index.depth() == layer_number)
+                .map(|(index, hash)| (*index, *hash))
+                .collect();
+
+            for (&index, &hash) in layer.iter() {
+                let mut parent = index;
+                parent.move_up();
+                if !nodes.contains_key(&parent) {
+                    let sibling = nodes.get(&index.sibling()).expect("sibling should exist");
+                    let parent_hash = Rpo256::merge(&index.build_node(hash, *sibling));
+                    nodes.insert(parent, parent_hash);
+                }
+            }
+        }
```

Comment:
How difficult would it be to create a separate `with_leaves()` constructor and use it here?

Reply:
It would be not hard, but since creation of the `PMT` from leaves is not a regular case, I think we have two options here:
- I can create full-fledged constructor with all necessary checks. It would work a little bit slower, but it can be used in other places.
- I can create "internal" constructor without additional checks (I'll just copy the code above) which will be used only here. That would be cleaner overall anyway 

So, the question is -- do we want to allow users to create `PMT` from leaves?

Reply:
Don't we need to perform all the checks anyway? There is no guarantee that the data we are deserializing was serialized correctly.

Reply:
Indeed, I missed this point. I'll create the proper constructor then.

Reply:
Should we return an error if a map with leaves also contain some internal node? Current implementation will just recalculate it with new value, but maybe it's not the best option.

Reply:
I'm thinking we should return an error, unless this introduces significant complications.